libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,223 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+   IBus telemetry should always run at 115200 baud rate.
+   It uses a half duplex protocol, so you need only to wire the TX pin to the telemetry port of the receiver
+   You can configure via your preferred GCS the serial port to send IBus telemetry data
+   You  need to initialize this object and then add sensors and set sensor measurements (about every 500ms - 1sec)
+*/
+#include "AP_IBus_Telem.h"
+#include "IBUSSensors.h"
+#include <AP_BattMonitor/AP_BattMonitor.h>
+#include <AP_Vehicle/AP_Vehicle.h>
+#include <AP_Arming/AP_Arming.h>
+
+
+extern const AP_HAL::HAL &hal;
+
+void AP_IBus_Telem::init()
+{
+    const AP_SerialManager &serial_manager = AP::serialmanager();
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_IBUS_Telem, 0)))
+    {
+        _port->set_options(_port->OPTION_HDPLEX);
+        _port->begin(AP_SERIALMANAGER_IBUS_TELEM_BAUD);
+        addSensors();
+
+        _initialized = true;
+    }
+}
+
+void AP_IBus_Telem::loop(void)
+{
+    //no need to run the loop if not initialized or no sensors added
+    if (!_initialized || _number_sensors == 0)
+    {
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
+    if (now - _last_send_measurements_time > UPDATE_MEASUREMENTS_TIMING)
+    {
+        setSensorMeasurements();
+    }
+    
+    // only process data already in our UART receive buffer
+    while (_port->available() > 0)
+    {
+        // only consider a new data package if we have not heard anything for >3ms
+        if (now - _last >= PROTOCOL_TIMEGAP)
+        {
+            _state = GET_LENGTH;
+        }
+        _last = now;
+
+        uint8_t v = _port->read();
```

Comment:
This is not the correct return type for `read`.  It's `int16_t` and you must check for `<0`

Reply:
v was for value :) but I will change it to c